File: 0054-spiral-matrix/0054-spiral-matrix.java
class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        int rows=matrix.length;
        int cols= matrix[0].length;
        int total = rows * cols;
        int[] result = new int[total];
        int top=0;
        int bottom =rows-1;
        int left=0;
        int right=cols-1;
        int i=0;
        while (i<total){
            for (int col=left; col<right+1;col++){
                if (i<total){
                    result[i]=matrix[top][col];
                    i+=1;
                }
                    
            }
            top+=1;
            for (int row=top; row<bottom+1;row++){
                if (i<total){
                    result[i]=matrix[row][right];
                    i+=1;
                }
                    
            }
            right-=1;

            for (int col=right; col>=left;col--){
                if (i<total){
                    result[i]=matrix[bottom][col];
                    i+=1;
                }
                    
            }
            bottom-=1;
            for (int row=bottom; row>=top;row--){
                if (i<total){
                    result[i]=matrix[row][left];
                    i+=1;
                }
                    
            }
            left+=1;


        }
        List<Integer> list = new ArrayList<>();
        for (int val : result) {
            list.add(val);
        }
        return list;

        
        
    }
}

File: 0054-spiral-matrix/0054-spiral-matrix.py
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        n=len(matrix)
        m=len(matrix[0])
        top=0
        left=0
        bot=n-1
        right=m-1
        ans=[]
        while top<=bot and left<=right:
            for i in range(left,right+1):
                ans.append(matrix[top][i])
            top+=1

            for i in range(top,bot+1):
                ans.append(matrix[i][right])
            right-=1

            if top<=bot:
                for i in range(right,left-1,-1):
                    ans.append(matrix[bot][i])
                bot-=1
            if left<=right:
                for i in range(bot,top-1,-1):
                    ans.append(matrix[i][left])
                left+=1
        return ans
                

        